plan_delegator/src/plan_delegator.cpp

```diff
@@ -167,6 +275,84 @@ namespace plan_delegator
         ROS_DEBUG_STREAM("adjusted_end_dist:" << adjusted_end_dist);
         SET_MANEUVER_PROPERTY(maneuver, start_dist, adjusted_start_dist);
         SET_MANEUVER_PROPERTY(maneuver, end_dist, adjusted_end_dist);
+
+        // Get the lanelets crossed by the updated maneuver (considers full route; not just shortest path)
+        std::vector<lanelet::ConstLanelet> adjusted_crossed_lanelets = wm_->getLaneletsBetween(adjusted_start_dist, adjusted_end_dist, false, false);
+        
+        if (adjusted_crossed_lanelets.size() == 0) {
+            throw std::invalid_argument("The adjusted maneuver does not cross any lanelets going from: " + std::to_string(adjusted_start_dist) + " to " + std::to_string(adjusted_end_dist));
+        }
+
+        // Update maneuver-specific lanelet ID parameters
+        // Note: Assumes that the maneuver start and end distances are adjusted by a distance less than the length of a lanelet. 
+        if(maneuver.type == cav_msgs::Maneuver::LANE_FOLLOWING) 
+        {
+            // Note: Assumes that strategic plugins have planned lane following maneuvers far enough into their specified 
+            //       ending lanelet that the maneuver ending lanelet remains unchanged.
+
+            // Obtain the original starting lanelet from the maneuver's lane_ids
+            lanelet::Id original_starting_lanelet_id = std::stoi(maneuver.lane_following_maneuver.lane_ids.front());
+            lanelet::ConstLanelet original_starting_lanelet = wm_->getMap()->laneletLayer.get(original_starting_lanelet_id);
+
+            // Check whether the lanelet prior to the original starting lanelet is crossed by the updated maneuver
+            for(auto lanelet : adjusted_crossed_lanelets) {
+                auto routing_relation = wm_->getMapRoutingGraph()->routingRelation(lanelet, original_starting_lanelet);
+
+                // Lanelet prior to the original starting lanelet is crossed by the updated maneuver, so it is added to the beginning of lane_ids
+                if (routing_relation == lanelet::routing::RelationType::Successor) {
+                    ROS_DEBUG_STREAM("Lanelet " << lanelet.id() << " added to front of maneuver's lane_ids");
+                    maneuver.lane_following_maneuver.lane_ids.insert(maneuver.lane_following_maneuver.lane_ids.begin(), std::to_string(lanelet.id()));
```

Comment:
I think as a final step the lane_ids need to be sorted by downtrack distance. 
It doesn't specify it in the message spec but I think somewhere I've seen an assumption that this list is ordered.

Reply:
It is assumed that strategic plugins populate the lane_ids array with lanelet ids in order of increasing downtrack distance, so the first element corresponds to the starting lanelet, and the last element corresponds to the ending lanelet. I have added comments to the code to state this.